cdc/processor/sinkmanager/table_sink_worker.go

```diff
@@ -58,223 +59,332 @@ func newSinkWorker(
-			appendEventsAndRecordCurrentSize := func() error {
-				size, err := w.appendEventsToTableSink(task, events)
-				if err != nil {
-					return errors.Trace(err)
-				}
-				log.Debug("append events to table sink",
+	// Used to record the last written position.
+	// We need to use it to update the lower bound of the table sink.
+	var lastPos engine.Position
+	currentCommitTs := uint64(0)
+	lastTimeCommitTs := uint64(0)
+	currentTotalSize := uint64(0)
+	batchID := uint64(1)
+
+	// Two functions to simplify the code.
+	// It captures some variables in the outer scope.
+	appendEventsAndRecordCurrentSize := func(commitTs model.Ts) error {
+		i := sort.Search(len(events), func(i int) bool {
+			return events[i].CRTs > commitTs
+		})
+		resolvedEvents := events[:i]
+		if len(resolvedEvents) == 0 {
+			return nil
+		}
+		size, err := w.appendEventsToTableSink(task, resolvedEvents)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		currentTotalSize += size
+		events = append(make([]*model.PolymorphicEvent, 0, len(events[i:])), events[i:]...)
+		return nil
+	}
+	advanceTableSinkAndResetCurrentSizeWithBatchID := func(commitTs model.Ts) error {
+		err := w.advanceTableSinkWithBatchID(task, commitTs, currentTotalSize, batchID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		currentTotalSize = 0
+		return nil
+	}
+	advanceTableSinkAndResetCurrentSize := func(commitTs model.Ts) error {
+		err := w.advanceTableSink(task, commitTs, currentTotalSize)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		currentTotalSize = 0
+		return nil
+	}
+
+	// lowerBound and upperBound are both closed intervals.
+	iter := engine.NewMountedEventIter(
+		w.sortEngine.FetchByTable(task.tableID, lowerBound, upperBound),
+		w.mg, 256)
+	defer func() {
+		if err := iter.Close(); err != nil {
+			log.Error("Sink worker fails to close iterator",
+				zap.String("namespace", w.changefeedID.Namespace),
+				zap.String("changefeed", w.changefeedID.ID),
+				zap.Int64("tableID", task.tableID),
+				zap.Error(err))
+		}
+
+		log.Debug("Sink worker handle task finished",
+			zap.String("namespace", w.changefeedID.Namespace),
+			zap.String("changefeed", w.changefeedID.ID),
+			zap.Int64("tableID", task.tableID),
+			zap.Any("lowerBound", lowerBound),
+			zap.Any("upperBound", upperBound),
+			zap.Bool("splitTxn", w.splitTxn),
+		)
+	}()
+	for !task.isCanceled() {
+		e, pos, err := iter.Next(ctx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		// There is no more data.
+		if e == nil {
+			log.Debug("No more data",
+				zap.String("namespace", w.changefeedID.Namespace),
+				zap.String("changefeed", w.changefeedID.ID),
+				zap.Int64("tableID", task.tableID),
+				zap.Any("lowerBound", lowerBound),
+				zap.Any("upperBound", upperBound),
+				zap.Uint64("currentCommitTs", currentCommitTs),
+				zap.Uint64("lastTimeCommitTs", lastTimeCommitTs),
+				zap.Uint64("currentTotalSize", currentTotalSize),
+				zap.Bool("splitTxn", w.splitTxn),
+			)
+			break
+		}
```

Comment:
add some comments here to explain why e.row= nil

Reply:
Added